server/login.go

```diff
@@ -21,12 +21,21 @@ var ErrInvalidUsernameOrPassword = api.Wrap(
 type LoginRequest struct {
 	Email    string `form:"email"`
 	Password string `form:"password"`
+
+	RedirectURI string `form:"redirect_uri"`
+}
+
+type LoginResponse struct {
+	RedirectURI string `form:"redirect_uri"`
 }
 
 func HandleLogin(ctx context.Context, req *http.Request) api.Response {
+	L := lib.GetLogger(ctx)
 	stores := models.GetStores(ctx)
 
 	// If the user is already logged in, redirect them to their user page.
+	// This gets a little bit hairy, because we need to catch invalid sessions and expired tokens,
+	// otherwise the login page 404s which isn't really all that helpful.
```

Comment:
I feel like the places where it's not desired behaviour is rather small tbh, and it can be worked around in those places. It's basically just when dealing with user models.

Reply:
I'm thinking e.g. feed page has potential to hit lots of RecordNotFounds as it expands objects but that's not a 404

But food for thought!